test/common/bidi.py

```diff
@@ -0,0 +1,438 @@
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close(self) -> None:
+        assert self.bidi_session is not None
+        log_proto.debug("cleaning up webdriver")
+
+        self.task_reader.cancel()
+        del self.task_reader
+        await self.ws.close()
+        await self.close_bidi_session()
+        self.bidi_session.process.terminate()
+        await self.bidi_session.process.wait()
+        self.bidi_session = None
+        await self.http_session.close()
+
+    def ws_done_callback(self, future):
+        for fut in self.pending_commands.values():
+            if not fut.done():
+                fut.set_exception(WebdriverError("websocket closed"))
+        if not future.cancelled():
+            log_proto.error("ws_reader crashed: %r", future.result())
+
+    async def start_session(self) -> None:
+        self.http_session = aiohttp.ClientSession(raise_for_status=True)
+        await self.start_bidi_session()
+        assert self.bidi_session
+        self.ws = await self.http_session.ws_connect(self.bidi_session.ws_url)
+        self.task_reader = asyncio.create_task(self.ws_reader(self.ws), name="bidi_reader")
+        self.task_reader.add_done_callback(self.ws_done_callback)
+
+        await self.bidi("session.subscribe", events=[
+            "log.entryAdded", "browsingContext.domContentLoaded",
+        ])
+
+        # wait for browser to initialize default context
+        for _ in range(10):
+            realms = (await self.bidi("script.getRealms"))["realms"]
+            if len(realms) > 0:
+                self.top_context = realms[0]["context"]
+                self.context = self.top_context
+                break
+            await asyncio.sleep(0.5)
+        else:
+            raise WebdriverError("timed out waiting for default realm")
+
+    async def __aenter__(self):
+        await self.start_session()
+        return self
+
+    async def __aexit__(self, *_excinfo):
+        if self.bidi_session is not None:
+            await self.close()
+
+    async def ws_reader(self, ws: aiohttp.client.ClientWebSocketResponse) -> None:
+        async for msg in ws:
+            if msg.type == aiohttp.WSMsgType.TEXT:
+                data = json.loads(msg.data)
+                log_proto.debug("ws TEXT → %r", data)
+                if "id" in data and data["id"] in self.pending_commands:
+                    log_proto.debug("ws_reader: resolving pending command %i", data["id"])
+                    if data["type"] == "success":
+                        if data["result"].get("type") == "exception":
+                            self.pending_commands[data["id"]].set_exception(
+                                WebdriverError(data["result"]["exceptionDetails"]["text"]))
+                        else:
+                            self.pending_commands[data["id"]].set_result(data["result"])
+                    elif data["type"] == "exception":
```

Comment:
What's the difference between `data.result.type == 'exception'` and `data.type == 'exception'`?  Is one of those for sync calls and the other for when you wait on promises?

Reply:
I feel like broadly speaking it might be more useful if the toplevel decision in this function weren't the tag lookup but rather on the "type" field of the message.  Like whenever you get "success" or "exception" types then you can expect to have the tag, right, so something like:

```python
  if data.type in ['success', 'exception']:
     pending = self.pending.get(data.get('id'))
```

and then for `type == 'event'` you should never have a tag, right?

Then you could unify your "unknown response type" blocks for the "with tag" and "without tag" cases.  I don't think it's actually important that this exception gets raised in the context of the tagged call, since it's a "should never happen thing" (and indeed, I don't expect at all to see a 'tag' on these unknown message types)...  You could even argue that the new unexpected message type could be something like a progress report type notification on an operation with a given tag, in which case returning an exception for that pending operation is outright wrong...

Reply:
> What's the difference between data.result.type == 'exception' and data.type == 'exception'

I think the first is when you do a valid BiDi  command which fails (like, a non-existing selector), and the other is when you send an invalid BiDi command (wrong argument or type or so).

I'd rather conceptually group the "response to request" (with "id") vs. "event" handling. When I do the `in ['success', 'exception']:` at the top, I have to assert that there is an ID (which isn't bad, but "meh"), and then repeat all the individual cases again, which is redundant and easy to get out of sync.